### ISSUE TITLE
NaN in eval loss

### ISSUE THREAD

**Issue by ultor1996** [McGill-NLP/llm2vec#174]
Hi, 
I am using the run_mntp.py script to riuna  mntp trainign on a model. When I do that I encounter an issue, my training loss goes down howver my eval loss is NaN.  You can see it in the file below. PLease let me know any insights into this.

[training.log](https://github.com/user-attachments/files/22006806/training.log)

**Comment by meaningful96** [comment]
Hi,
I’m seeing the same behavior: training loss decreases, but eval loss becomes `NaN`.

```python

word_embeddings": false,
  "torch_dtype": "bfloat16",
  "transformers_version": "4.44.2",
  "use_cache": true,
  "vocab_size": 128256
}

[INFO|trainer.py:3824] 2025-10-22 21:20:22,700 >>   Batch size = 24
{'eval_loss': nan, 'eval_accuracy': 0.0, 'eval_runtime': 17.3192, 'eval_samples_per_second': 28.812, 'eval_steps_per_second': 0.346, 'epoch': 0.08}                         
  3%|███▍                                                                                                                             | 200/7413 [32:53<19:20:33,  9.65s/it10/22/2025 21:20:40 - INFO - __main__ - Saving model checkpoint to output/mntp/Meta-Llama-3-8B-Instruct/checkpoint-200                                                       
[INFO|configuration_utils.py:733] 2025-10-22 21:20:40,816 >> loading configuration file config.json from cache at 
[INFO|configuration_utils.py:800] 2025-10-22 21:20:40,818 >> Model config LlamaConfig {
  "architectures": [
    "LlamaForCausalLM"
  ],
  "attention_bias": false,
  "attention_dropout": 0.0,
  "bos_token_id": 128000,
  "eos_token_id": 128009,
  "hidden_act": "silu",
  "hidden_size": 4096,
  "initializer_range": 0.02,
  "intermediate_size": 14336,
  "max_position_embeddings": 8192,
  "mlp_bias": false,
  "model_type": "llama",
  "num_attention_heads": 32,
  "num_hidden_layers": 32,
  "num_key_value_heads": 8,
  "pretraining_tp": 1,
  "rms_norm_eps": 1e-05,
  "rope_scaling": null,
  "rope_theta": 500000.0,
  "tie_word_embeddings": false,
  "torch_dtype": "bfloat16",
  "transformers_version": "4.44.2",
  "use_cache": true,
  "vocab_size": 128256
}

[INFO|tokenization_utils_base.py:2684] 2025-10-22 21:20:41,160 >> tokenizer config file saved in output/mntp/Meta-Llama-3-8B-Instruct/checkpoint-200/tokenizer_config.json
[INFO|tokenization_utils_base.py:2693] 2025-10-22 21:20:41,160 >> Special tokens file saved in output/mntp/Meta-Llama-3-8B-Instruct/checkpoint-200/special_tokens_map.json

  with device_autocast_ctx, torch.cpu.amp.autocast(**cpu_autocast_kwargs), recompute_context:  # type: ignore[attr-defined]
  4%|█████▏                                                                                                                           | 300/7413 [49:05<19:09:19,  9.69s/it]
[INFO|trainer.py:3819] 2025-10-22 21:36:51,546 >> 
***** Running Evaluation *****
[INFO|trainer.py:3821] 2025-10-22 21:36:51,546 >>   Num examples = 499
[INFO|trainer.py:3824] 2025-10-22 21:36:51,547 >>   Batch size = 24
{'eval_loss': nan, 'eval_accuracy': 0.0, 'eval_runtime': 17.3223, 'eval_samples_per_second': 28.807, 'eval_steps_per_second': 0.346, 'epoch': 0.12}                         
  with device_autocast_ctx, torch.cpu.amp.autocast(**cpu_autocast_kwargs), recompute_context:  # type: ignore[attr-defined]
  5%|██████▊                                                                                                                        | 400/7413 [1:05:28<18:47:48,  9.65s/it]
[INFO|trainer.py:3819] 2025-10-22 21:53:14,538 >> 
***** Running Evaluation *****
[INFO|trainer.py:3821] 2025-10-22 21:53:14,538 >>   Num examples = 499
[INFO|trainer.py:3824] 2025-10-22 21:53:14,539 >>   Batch size = 24
{'eval_loss': nan, 'eval_accuracy': 0.0, 'eval_runtime': 17.3012, 'eval_samples_per_second': 28.842, 'eval_steps_per_second': 0.347, 'epoch': 0.16}                         
  5%|██████▊                                                                                                                        | 400/7413 [1:05:45<18:47:48,  9.65s/it10/22/2025 21:53:31 - INFO - __main__ - Saving model checkpoint to output/mntp/Meta-Llama-3-8B-Instruct/checkpoint-400                                                       
[INFO|configuration_utils.py:733] 2025-10-22 21:53:32,660 >> loading configuration file config.json from cache at \
[INFO|configuration_utils.py:800] 2025-10-22 21:53:32,661 >> Model config LlamaConfig {
  "architectures": [
    "LlamaForCausalLM"
  ],
  "attention_bias": false,
 ```

**Comment by meaningful96** [comment]
Hi, 

I ran into the same issue and resolved it by adjusting precision per model family: on my setup Qwen trains stably with `float16` (fp16), but Gemma, Llama, and Mistral consistently produce NaN loss under fp16 and only stabilize with `bfloat16` (bf16) or `float32`; switching my config to disable fp16 and enable bf16 (or loading the model with torch_dtype set to bfloat16/float32) eliminated the `NaN` for those models while keeping Qwen on fp16 without problems—hope this helps.
